src/ripple/rpc/impl/ShardArchiveHandler.cpp

```diff
@@ -61,24 +94,156 @@ ShardArchiveHandler::hasInstance()
     , validate_(validate)
     , timer_(app_.getIOService())
     , process_(false)
-    , j_(app.journal("ShardArchiveHandler"))
 {
     assert(app_.getShardStore());
+
+    auto setup = setup_DatabaseCon(app_.config());
+    setup.dataDir = stateDir_;
+
+    create_directory(stateDir_);
+
+    conn_ = std::make_unique<DatabaseCon>(
+        stateDir_,
+        "stateDatabase",
+        DatabasePragma,
+        DatabaseInit);
+
+    auto db = conn_->checkoutDb();
+
+    // Always use named variables with 'soci::use'.
+    // Avoid temporaries.
+    int v = int(validate_);
+
+    *db << "INSERT INTO MetaState VALUES (\"validate\", :val);",
+        soci::use(v);
+
+    *db << "INSERT INTO MetaState VALUES (\"paused\", 0);";
+
+    (void)tag;
+}
+
+ShardArchiveHandler::ShardArchiveHandler(WithRecovery const& tag,
+                                         Application& app, bool & invokeStart)
+    : app_(app)
+    , j_(app.journal("ShardArchiveHandler"))
+    , downloader_(std::make_shared<DatabaseDownloader>(
+        app_.getIOService(), j_, app_.config()))
+    , downloadDir_(getDownloadDirectory(app.config()))
+    , stateDir_(getStateDirectory(app.config()))
+    , timer_(app_.getIOService())
+    , process_(false)
+{
+    auto setup = setup_DatabaseCon(app_.config());
+    setup.dataDir = stateDir_;
+
+    using namespace boost::filesystem;
+
+    assert(exists(stateDir_.string() + "/stateDatabase") &&
+           is_regular_file(stateDir_.string() + "/stateDatabase"));
+
+    conn_ = std::make_unique<DatabaseCon>(
+        stateDir_,
+        "stateDatabase",
+        DatabasePragma,
+        DatabaseInit);
+
+    auto db = conn_->checkoutDb();
+
+    int paused;
+    int validate;
+
+    *db << "SELECT Value FROM MetaState WHERE Key=\"paused\";",
+        soci::into(paused);
+
+    *db << "SELECT Value FROM MetaState WHERE Key=\"validate\";",
+        soci::into(validate);
+
+    validate_ = validate;
+
+    soci::rowset<soci::row> rs = (db->prepare
+        << "SELECT * FROM State;");
+
+    for (auto it = rs.begin(); it != rs.end(); ++it)
+    {
+        parsedURL url;
+
+        if (!parseUrl(url, it->get<std::string>(1)))
+        {
+            JLOG(j_.error()) << "Failed to parse url: "
+                << it->get<std::string>(1);
+
+            continue;
+        }
+
+        uint64_t index = 0;
+
+        switch(it->get_properties(0).get_data_type())
+        {
+            case soci::dt_double:
+                index = it->get<double>(0);
+                break;
+            case soci::dt_integer:
+                index = it->get<int>(0);
+                break;
+            case soci::dt_long_long:
+                index = it->get<long long>(0);
+                break;
+            case soci::dt_unsigned_long_long:
+                index = it->get<unsigned long long>(0);
+                break;
+            default:
+                JLOG(j_.error()) << "Failed to parse shard index.";
+                continue;
+        }
```

Comment:
This switch seems unnecessary as we should never be reading or writing anything other than a INTEGER for shardIndex. Consider omitting the switch with something like.
```
add(it->get<int>(0), std::move(url), std::lock_guard<std::mutex>(m_));
```

Reply:
During verification testing I was seeing the numeric types vary randomly. I will try again and make this change if its not happening anymore

Reply:
Done